Use the cookie_store package for state management

egefeyzioglu opened the pull request:
Closes #1, closes #2

Also actually checks who the cookie belongs to before sending it along, which is critical for security.

egefeyzioglu committed:
Use the cookie_store package to manage the session

egefeyzioglu committed:
Use non-broken version of cookie_store

egefeyzioglu committed:
Add test for new cookie store

egefeyzioglu committed:
dart fix fixes

egefeyzioglu committed:
Version bump

lamnhan066 commented on lib/src/http_session.dart:
```diff
@@ -4,21 +4,23 @@ import 'dart:typed_data';
   static final shared = HttpSession();
 
-  /// Get current headers value
-  Map<String, String> get headers => _headers;
```
I think we should keep this getter to be compatible with the older version.

egefeyzioglu replied:
It's not really possible to have a useful presistent `headers` field anymore since not all cookies are returned for all requests. Exposing all the cookies through a field like that would encourage using it unsafely and accessing cookies that aren't supposed to be returned with a request.

lamnhan066 replied:
That makes sense. It should be deprecated.

lamnhan066 commented on lib/src/http_session.dart:
```diff
@@ -61,34 +67,34 @@ class HttpSession implements IOClient {
     Encoding? encoding,
   }) async {
-    return _updateResponse(await _httpDelegate.delete(
+    Map<String, String> headers0 = {
+      "Cookie": _getCookieHeader(url.host, url.path)
+    };
+    headers0.addAll(headers ?? {});
```
Do you think this code should be rewrote as a function so we can reuse it and also easier to maintain?

egefeyzioglu replied:
I agree but I couldn't come up with a good way to do this, any ideas?

lamnhan066 replied:
We can do something like this:

``` dart
  Map<String, String> _getHeaders(Uri url, Map<String, String> headers) {
    Map<String, String> tempHeaders = {
      "Cookie": _getCookieHeader(url.host, url.path)
    };
    tempHeaders.addAll(headers ?? {});
    return tempHeaders;
  }
```

Then we can pass the `_getHeaders(url, headers)` to all `headers` parameters.

egefeyzioglu replied:
I ended up rewriting that bit when trying to get it to support redirects ([a2ffb8d](https://github.com/lamnhan066/http_session/pull/5/commits/a2ffb8d3419d04ff2ef405990b92b937495431ee))

Lmk what you think please

lamnhan066 replied:
Can we use `_sendRequest` for this `delete` method?

egefeyzioglu replied:
Oops, fixed

lamnhan066 reviewed (commented):
I have left some comments for the new code

egefeyzioglu committed:
Don't miss cookies during redirection

egefeyzioglu committed:
Remove unused import

egefeyzioglu committed:
Provide information about the redirects when throwing

egefeyzioglu committed:
Depend on the newer version of cookie_store

egefeyzioglu committed:
Use adhoc server instead of an external one

egefeyzioglu committed:
Remove unused import

egefeyzioglu committed:
Update change log

lamnhan066 commented on lib/src/http_session.dart:
```diff
@@ -1,24 +1,29 @@
 
-  /// Get current headers value
-  Map<String, String> get headers => _headers;
+  int maxRedirects = 15;
```
I think we should add this variable to the `HttpSession` constructor so we can set it directly.

egefeyzioglu replied:
Oh I meant to do that my bad

egefeyzioglu replied:
Done

lamnhan066 commented on lib/src/http_session.dart:
```diff
@@ -61,34 +70,83 @@ class HttpSession implements IOClient {
     ));
   }
 
+  Future<http.Response> _sendRequest(String method, Uri url, int timeToLive,
```
I get a little bit confused when I see 'timeToLive', do you think this is the right word to describe this behavior?

egefeyzioglu replied:
Replaced it with `numRecurseLeft`

egefeyzioglu committed:
Allow setting maxRedirects from constructor

egefeyzioglu committed:
Use _sendRequest for delete() as well

egefeyzioglu committed:
Better variable naming

Change timeToLive to numRecurseLeft in _sendRequest

lamnhan066 commented on lib/src/http_session.dart:
```diff
@@ -139,6 +120,17 @@ class HttpSession implements IOClient {
+    Object? body,
+    Encoding? encoding,
+  }) async {
+    return _sendRequest("POST", url, maxRedirects,
```
I think this should be "DELETE".

egefeyzioglu replied:
Oh wow I'm dumb yes it does need to be DELETE

egefeyzioglu replied:
I'm fixing this also I'm adding tests to make sure nothing like this gets missed in the future

egefeyzioglu replied:
Done

egefeyzioglu committed:
Fix typo in HTTP method name

egefeyzioglu committed:
Add more tests

- Test that we send the correct verbs
- Test that we're not sending cookies to the wrong domains

egefeyzioglu committed:
Test HEAD method as well

lamnhan066 commented:
@egefeyzioglu I noticed that you haven’t solved #2 yet. How do you think about the solution that I mentioned in #2?

egefeyzioglu committed:
Resolve #1

Use the solution recommended by lamnhan066 in the issue comments

egefeyzioglu commented:
Oh I thought you'd fixed that already in the branch I forked from, my bad, applied your solution now

egefeyzioglu committed:
Add better test for .head()

egefeyzioglu committed:
Update README to match new API

lamnhan066 commented:
@egefeyzioglu Should we export the `cookie_store` so the users don't need to add it to the `dependencies` when they want to get `cookieStore`?

egefeyzioglu committed:
Export cookie_store

egefeyzioglu commented:
Good idea, done!

egefeyzioglu committed:
Update changelog

lamnhan066 commented:
I think we have completed the review on the `http_session` side. I cannot review the `cookie_store` so I'll merge this PR when you confirm that the `cookie_store` is stable.

egefeyzioglu commented:
Thanks very much! I can confirm that `cookie_store` is stable
